File: ios/runanywhere-sdks-latest/examples/ios/RunAnywhereAI/RunAnywhereAI/Core/Models/AppTypes.swift
//
//  AppTypes.swift
//  RunAnywhereAI
//
//  Essential app types - using SDK types directly
//

import Foundation
import RunAnywhere

// MARK: - System Device Info

struct SystemDeviceInfo {
    let modelName: String
    let chipName: String
    let totalMemory: Int64
    let availableMemory: Int64
    let neuralEngineAvailable: Bool
    let osVersion: String
    let appVersion: String

    init(
        modelName: String = "",
        chipName: String = "",
        totalMemory: Int64 = 0,
        availableMemory: Int64 = 0,
        neuralEngineAvailable: Bool = false,
        osVersion: String = "",
        appVersion: String = ""
    ) {
        self.modelName = modelName
        self.chipName = chipName
        self.totalMemory = totalMemory
        self.availableMemory = availableMemory
        self.neuralEngineAvailable = neuralEngineAvailable
        self.osVersion = osVersion
        self.appVersion = appVersion
    }
}

// MARK: - Helper Extensions

extension Int64 {
    var formattedFileSize: String {
        let formatter = ByteCountFormatter()
        formatter.allowedUnits = [.useBytes, .useKB, .useMB, .useGB]
        formatter.countStyle = .file
        return formatter.string(fromByteCount: self)
    }
}

File: ios/runanywhere-sdks-latest/examples/flutter/RunAnywhereAI/lib/core/models/app_types.dart
// App Types (mirroring iOS AppTypes.swift)
//
// Contains core data models used throughout the app.

/// System device information for displaying hardware capabilities
class SystemDeviceInfo {
  final String modelName;
  final String chipName;
  final int totalMemory;
  final int availableMemory;
  final bool neuralEngineAvailable;
  final String osVersion;
  final String appVersion;

  const SystemDeviceInfo({
    this.modelName = '',
    this.chipName = '',
    this.totalMemory = 0,
    this.availableMemory = 0,
    this.neuralEngineAvailable = false,
    this.osVersion = '',
    this.appVersion = '',
  });

  SystemDeviceInfo copyWith({
    String? modelName,
    String? chipName,
    int? totalMemory,
    int? availableMemory,
    bool? neuralEngineAvailable,
    String? osVersion,
    String? appVersion,
  }) {
    return SystemDeviceInfo(
      modelName: modelName ?? this.modelName,
      chipName: chipName ?? this.chipName,
      totalMemory: totalMemory ?? this.totalMemory,
      availableMemory: availableMemory ?? this.availableMemory,
      neuralEngineAvailable:
          neuralEngineAvailable ?? this.neuralEngineAvailable,
      osVersion: osVersion ?? this.osVersion,
      appVersion: appVersion ?? this.appVersion,
    );
  }
}

/// Extension for formatting file sizes
extension FileSizeFormatter on int {
  /// Formats bytes into human-readable file size string
  String get formattedFileSize {
    const units = ['B', 'KB', 'MB', 'GB', 'TB'];
    double size = toDouble();
    int unitIndex = 0;

    while (size >= 1024 && unitIndex < units.length - 1) {
      size /= 1024;
      unitIndex++;
    }

    if (unitIndex == 0) {
      return '${size.toInt()} ${units[unitIndex]}';
    }
    return '${size.toStringAsFixed(1)} ${units[unitIndex]}';
  }
}

// MessageRole is now provided by the RunAnywhere SDK
// import 'package:runanywhere/runanywhere.dart' show MessageRole;

/// Completion status for message generation
enum CompletionStatus {
  complete,
  interrupted,
  failed,
  timeout,
}

/// Generation mode for LLM inference
enum GenerationMode {
  streaming,
  nonStreaming,
}

/// Routing policy for model selection
enum RoutingPolicy {
  automatic,
  deviceOnly,
  preferDevice,
  preferCloud,
}

// ModelLoadState is now provided by the RunAnywhere SDK
// import 'package:runanywhere/runanywhere.dart' show ModelLoadState;
// Use AppModelLoadState for app-specific states to avoid conflict
enum AppModelLoadState {
  notLoaded,
  loading,
  loaded,
  failed,
}

/// Voice session state for voice assistant
enum VoiceSessionState {
  disconnected,
  connecting,
  connected,
  listening,
  processing,
  speaking,
  error,
}
